Metric spaces (#1162)
UniMath/agda-unimath: src/metric-spaces/premetric-structures.lagda.md

Reviewer: This file is still quite long. I think it would still be a benefit to refactor out some definitions + properties to new files, since many of the concepts are independent of eachother

Author: OK. I'll split it up, with a separate module for each property.

Author: do you like it better now?

Reviewer: Great work, this looks much neater now!

# Premetric structures on types

```agda
module metric-spaces.premetric-structures where
```

<details><summary>Imports</summary>

```agda
open import elementary-number-theory.positive-rational-numbers

open import foundation.binary-relations
open import foundation.dependent-pair-types
open import foundation.empty-types
open import foundation.equivalences
open import foundation.existential-quantification
open import foundation.function-extensionality
open import foundation.function-types
open import foundation.fundamental-theorem-of-identity-types
open import foundation.identity-types
open import foundation.logical-equivalences
open import foundation.negation
open import foundation.propositional-extensionality
open import foundation.propositions
open import foundation.sets
open import foundation.subtypes
open import foundation.torsorial-type-families
open import foundation.transport-along-identifications
open import foundation.univalence
open import foundation.universe-levels
```

</details>

## Idea

> A {{#concept "premetric" Disambiguation="structure on a type" Agda=Premetric}}
> is a type family of
> [proposition-valued binary relations](foundation.binary-relations.md) indexed by
> the
> [positive rational numbers](elementary-number-theory.positive-rational-numbers.md).
> 
> This follows Definition 4.5.2 from {{#cite Booij2020PhD}}.
> 
> In the classical interpretation, for a premetric `B` on a type `A` and some
> positive rational number `d : ℚ⁺`, `B d x y` holds for some points `(x y : A)`
> if the distance between `x` and `y` is bounded above by `d`.

Reviewer: ```suggestion
A {{#concept "premetric" Disambiguation="structure on a type" Agda=Premetric}}
is a type family of
[proposition-valued binary relations](foundation.binary-relations.md) indexed by
the
[positive rational numbers](elementary-number-theory.positive-rational-numbers.md).

Given a premetric `B` on `A` and some
positive rational number `d : ℚ⁺` such that `B d x y` holds for some pair of points `x y : A`, we interpret `d` as an {{#concept "upper bound" Disambiguation="on distance in with respect to a premetric structure"}}
on the distance between `x` and `y` with respect to the premetric.
```

  is-torsorial-Eq-Premetric =
    ( N , refl-Eq-Premetric) ,
    ( λ (N' , e) →
      eq-type-subtype
        ( Eq-prop-Premetric)
        ( eq-Eq-Premetric N' e))

  is-fiberwise-equiv-Eq-eq-Premetric :
    (N' : Premetric l2 A) → is-equiv (Eq-eq-Premetric N')
  is-fiberwise-equiv-Eq-eq-Premetric =
    fundamental-theorem-id is-torsorial-Eq-Premetric Eq-eq-Premetric

  equiv-Eq-eq-Premetric :
    (N' : Premetric l2 A) → (N ＝ N') ≃ (Eq-Premetric N')
  equiv-Eq-eq-Premetric N' =
    Eq-eq-Premetric N' , is-fiberwise-equiv-Eq-eq-Premetric N'
```

### Characterization of the transport of premetric structures along equality of types

```agda
module _
  {l1 l2 : Level} (A : UU l1)
  where

  eq-map-eq-tr-Premetric :
    (B : UU l1) (e : A ＝ B) (S : Premetric l2 A) →
    Eq-Premetric S (λ d x y → tr (Premetric l2) e S d (map-eq e x) (map-eq e y))
  eq-map-eq-tr-Premetric .A refl S = refl-Eq-Premetric S

  eq-map-inv-eq-tr-Premetric :
    (B : UU l1) (e : A ＝ B) (S : Premetric l2 A) →
    Eq-Premetric
      (tr (Premetric l2) e S)
      (λ d x y → S d (map-inv-eq e x) (map-inv-eq e y))
  eq-map-inv-eq-tr-Premetric .A refl S = refl-Eq-Premetric S
```

> ## References
> 
> {{#bibliography}}

Reviewer: ```suggestion
## References

Our definition of a premetric follows Definition 4.5.2 from {{#cite Booij2020PhD}}.

{{#bibliography}}
```

open import foundation.propositions
open import foundation.sets
open import foundation.subtypes
open import foundation.torsorial-type-families
open import foundation.transport-along-identifications
open import foundation.univalence
open import foundation.universe-levels
```

</details>

## Idea

A {{#concept "premetric" Disambiguation="structure on a type" Agda=Premetric}}
is a type family of
[proposition-valued binary relations](foundation.binary-relations.md) indexed by
the
[positive rational numbers](elementary-number-theory.positive-rational-numbers.md).

This follows Definition 4.5.2 from {{#cite Booij2020PhD}}.

## Definitions

### Premetric structures

```agda
module _
  {l1 : Level} (l2 : Level) (A : UU l1)
  where

  Premetric : UU (l1 ⊔ lsuc l2)
  Premetric = ℚ⁺ → Relation-Prop l2 A
```

### Closeness relation in a premetric

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  where

  neighborhood-Premetric : ℚ⁺ → A → A → UU l2
  neighborhood-Premetric d = type-Relation-Prop (B d)

  is-prop-neighborhood-Premetric :
    (d : ℚ⁺) (x y : A) → is-prop (neighborhood-Premetric d x y)
  is-prop-neighborhood-Premetric d = is-prop-type-Relation-Prop (B d)
```

> ### Two elements `x` and `y` are indistinguishable in a premetric if `x` and `y` are `d`-neighbours for any positive rational `d`

Reviewer: Still the header shouldn't define the term

```md
### Indistinguishable elements with respect to a premetric

Two elements `x` and `y` are {{#concept "indistinguishable" Disambiguation="with respect to a premetric" Agda=...}} in a premetric if `x` and `y` are `d`-neighbours for any positive rational `d`.

```

Reviewer: oh, sorry, I was looking at an outdated version

Author: ok. I think I get it. I hadn't realize I could add prose or define new concept away from `## Idea` section.

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  (x y : A)
  where

  is-indistinguishable-prop-Premetric : Prop l2
  is-indistinguishable-prop-Premetric =
    Π-Prop ℚ⁺ (λ (d : ℚ⁺) → B d x y)

  is-indistinguishable-Premetric : UU l2
  is-indistinguishable-Premetric =
    type-Prop is-indistinguishable-prop-Premetric

  is-prop-is-indistinguishable-Premetric :
    is-prop is-indistinguishable-Premetric
  is-prop-is-indistinguishable-Premetric =
    is-prop-type-Prop is-indistinguishable-prop-Premetric
```

## Properties

> ### A premetric is reflexive is any element is indistinguishable from itself

Reviewer: move this to the definitions section and change the header

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  where

  is-reflexive-prop-Premetric : Prop (l1 ⊔ l2)
  is-reflexive-prop-Premetric =
    Π-Prop ℚ⁺ (is-reflexive-prop-Relation-Prop ∘ B)

  is-reflexive-Premetric : UU (l1 ⊔ l2)
  is-reflexive-Premetric = type-Prop is-reflexive-prop-Premetric

  is-prop-is-reflexive-Premetric : is-prop is-reflexive-Premetric
  is-prop-is-reflexive-Premetric =
    is-prop-type-Prop is-reflexive-prop-Premetric
```

### Indistiguishability in a reflexive premetric is reflexive

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  (R : is-reflexive-Premetric B)
  where

  is-reflexive-is-indistinguishable-reflexive-Premetric :
    is-reflexive (is-indistinguishable-Premetric B)
  is-reflexive-is-indistinguishable-reflexive-Premetric x d = R d x
```

### In a reflexive premetric, equal elements are indistinguishable

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  (H : is-reflexive-Premetric B)
  where

  indistinguishable-eq-reflexive-Premetric :
    {x y : A} → x ＝ y → is-indistinguishable-Premetric B x y
  indistinguishable-eq-reflexive-Premetric {x} {.x} refl d = H d x
```

> ### A premetric is symmetric if `d`-neighbourhoods are symmetric for all positive rational number `d`

Reviewer: move this to the definitions section and change the header as for indistinguishable elements

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  where

  is-symmetric-prop-Premetric : Prop (l1 ⊔ l2)
  is-symmetric-prop-Premetric =
    Π-Prop ℚ⁺ (is-symmetric-prop-Relation-Prop ∘ B)

  is-symmetric-Premetric : UU (l1 ⊔ l2)
  is-symmetric-Premetric = type-Prop is-symmetric-prop-Premetric

  is-prop-is-symmetric-Premetric : is-prop is-symmetric-Premetric
  is-prop-is-symmetric-Premetric =
    is-prop-type-Prop is-symmetric-prop-Premetric
```

### Indistiguishability in a symmetric premetric is symmetric

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  (S : is-symmetric-Premetric B)
  where

  is-symmetric-is-indistinguishable-is-symmetric-Premetric :
    is-symmetric (is-indistinguishable-Premetric B)
  is-symmetric-is-indistinguishable-is-symmetric-Premetric x y H d =
    S d x y (H d)
```

> ### A premetric is local if indistinguishability has propositional fibers

Reviewer: same here, move to definitions and change header + definition text

Reviewer: Also, I don't see why you discern between tightness and locality

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  where

  is-local-prop-Premetric : Prop (l1 ⊔ l2)
  is-local-prop-Premetric =
    Π-Prop A (is-prop-Prop ∘ Σ A ∘ is-indistinguishable-Premetric B)

  is-local-Premetric : UU (l1 ⊔ l2)
  is-local-Premetric = type-Prop is-local-prop-Premetric

  is-prop-is-local-Premetric : is-prop is-local-Premetric
  is-prop-is-local-Premetric = is-prop-type-Prop is-local-prop-Premetric
```

> ### In a local reflexive premetric, indistinguishability is torsorial

Reviewer: The more important conclusion is that indistinguishability characterizes equality imo

Reviewer: and the reader may not know what torsoriality means, so my suggestion would be more recognizable

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  (R : is-reflexive-Premetric B) (L : is-local-Premetric B) (x : A)
  where

  is-torsorial-indistinguishable-local-reflexive-Premetric :
    is-torsorial (is-indistinguishable-Premetric B x)
  is-torsorial-indistinguishable-local-reflexive-Premetric =
    is-proof-irrelevant-is-prop (L x) (x , λ d → R d x)

  is-fiberwise-equiv-indistinguishable-local-reflexive-Premetric :
    (y : A) → is-equiv (indistinguishable-eq-reflexive-Premetric B R)
  is-fiberwise-equiv-indistinguishable-local-reflexive-Premetric =
    fundamental-theorem-id
      ( is-torsorial-indistinguishable-local-reflexive-Premetric)
      ( λ y → indistinguishable-eq-reflexive-Premetric B R {x} {y})
```

> ### Any type equipped with a reflexive local premetric is a set

Reviewer: a good name for a reflexive local premetric is an _extensional_ premetric

Reviewer: you could create a new file about extensional premetrics in which you define reflexivity, and locality (and perhaps tightness, although I'm still not sure why you want that one around), and show properties derived from them. But then I would also be inclined to suggest that you create a separate file for reflexive premetrics

Author: >  I'm still not sure why you want that one around

It's a bit easier to prove than locality (e.g. to define the metric space of rationals or reals) and to work with (e.g. for subspaces / dependent products of metric spaces).

Reviewer: I'm not going to ask you to remove it, just keep in mind that it is not a homotopy coherent notion, much like injectivity for functions on types that are not sets. I suppose that's a useful lemma though, that tightness + your carrier type is a set implies local.

Author: tightness always implies locality (see `is-local-is-tight-Premetric`).

Author: Maybe you want something like "the discrete metric on a type is local if and only if it is a set"?

Author: see b2669541555ecaab855ddadf95c71d9d60c41ee5

Reviewer: very nice!

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  (R : is-reflexive-Premetric B) (L : is-local-Premetric B)
  where

  is-set-has-local-reflexive-Premetric : is-set A
  is-set-has-local-reflexive-Premetric x y =
    is-prop-is-equiv
      ( is-fiberwise-equiv-indistinguishable-local-reflexive-Premetric
        B
        R
        L
        x
        y)
      ( is-prop-is-indistinguishable-Premetric B x y)
```

> ### A premetric is tight if any two indistinguishable elements are equal

Reviewer: move to definitions + change header + add definition prose

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  where

  is-tight-Premetric : UU (l1 ⊔ l2)
  is-tight-Premetric =
    (x y : A) → is-indistinguishable-Premetric B x y → x ＝ y
```

### Any tight premetric is local

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  (T : is-tight-Premetric B)
  where

  is-local-is-tight-Premetric : is-local-Premetric B
  is-local-is-tight-Premetric x =
    is-prop-all-elements-equal
      ( λ (u , I) (v , J) →
        eq-type-subtype
          ( is-indistinguishable-prop-Premetric B x)
          ( inv (T x u I) ∙ T x v J))
```

### Any reflexive local premetric is tight

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  (R : is-reflexive-Premetric B) (L : is-local-Premetric B)
  where

  is-tight-is-local-reflexive-Premetric : is-tight-Premetric B
  is-tight-is-local-reflexive-Premetric x =
    ( map-inv-is-equiv) ∘
    ( is-fiberwise-equiv-indistinguishable-local-reflexive-Premetric B R L x)
```

> ### A premetric is monotonic if any `d₁`-neighbourhoods are `d₂`-neighbourhoods for `d₁ < d₂`

Author: How is it different from `is-reflexive-`, `is-symmetric-`, etc.?

> ### A premetric is monotonic if any `d₁`-neighbourhoods are `d₂`-neighbourhoods for `d₁ < d₂`

Reviewer: The difference is that we haven't refactored `is-reflexive` and `is-symmetric`

> ### A premetric is monotonic if any `d₁`-neighbourhoods are `d₂`-neighbourhoods for `d₁ < d₂`

Reviewer: Also, I believe in those cases the definitions don't come in/after a Property section

> ### A premetric is monotonic if any `d₁`-neighbourhoods are `d₂`-neighbourhoods for `d₁ < d₂`

Author: oh. So maybe I should just move all the `is-xxx-Premetric` definitions in the `## Definitions` section (in the sense of "definitions of some properties a premetric can have") and then in the `## Properties` section, have the different relations between them?

> ### A premetric is monotonic if any `d₁`-neighbourhoods are `d₂`-neighbourhoods for `d₁ < d₂`

Reviewer: that's certainly an improvement over the current version

> ### A premetric is monotonic if any `d₁`-neighbourhoods are `d₂`-neighbourhoods for `d₁ < d₂`

Author: Do you like te new version better?

  is-prop-Eq-Premetric : (N' : Premetric l2 A) → is-prop (Eq-Premetric N')
  is-prop-Eq-Premetric N' = is-prop-type-Prop (Eq-prop-Premetric N')

  refl-Eq-Premetric : Eq-Premetric N
  refl-Eq-Premetric d x y = id-iff

  Eq-eq-Premetric : (N' : Premetric l2 A) → (N ＝ N') → Eq-Premetric N'
  Eq-eq-Premetric .N refl = refl-Eq-Premetric

  eq-Eq-Premetric : (N' : Premetric l2 A) → Eq-Premetric N' → N ＝ N'
  eq-Eq-Premetric N' H =
    eq-htpy
      ( λ d →
        eq-htpy
        ( λ x →
          eq-htpy
          ( λ y →
            eq-iff' (N d x y) (N' d x y) (H d x y))))

  is-torsorial-Eq-Premetric : is-torsorial Eq-Premetric
  is-torsorial-Eq-Premetric =
    ( N , refl-Eq-Premetric) ,
    ( λ (N' , e) →
      eq-type-subtype
        ( Eq-prop-Premetric)
        ( eq-Eq-Premetric N' e))

  is-fiberwise-equiv-Eq-eq-Premetric :
    (N' : Premetric l2 A) → is-equiv (Eq-eq-Premetric N')
  is-fiberwise-equiv-Eq-eq-Premetric =
    fundamental-theorem-id is-torsorial-Eq-Premetric Eq-eq-Premetric

  equiv-Eq-eq-Premetric :
    (N' : Premetric l2 A) → (N ＝ N') ≃ (Eq-Premetric N')
  equiv-Eq-eq-Premetric N' =
    Eq-eq-Premetric N' , is-fiberwise-equiv-Eq-eq-Premetric N'
```

> ### Characterisation of the transport of premetric structures along equality of types

Author: oh. sorry. I'll change `neighbourhood` into `neighborhood` too.

```agda
module _
  {l1 : Level} (l2 : Level) (A : UU l1)
  where

  Premetric : UU (l1 ⊔ lsuc l2)
  Premetric = ℚ⁺ → Relation-Prop l2 A
```

### Closeness relation in a premetric

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  where

  neighborhood-Premetric : ℚ⁺ → A → A → UU l2
  neighborhood-Premetric d = type-Relation-Prop (B d)

  is-prop-neighborhood-Premetric :
    (d : ℚ⁺) (x y : A) → is-prop (neighborhood-Premetric d x y)
  is-prop-neighborhood-Premetric d = is-prop-type-Relation-Prop (B d)
```

### Two points `x` and `y` are indistinguishable in a premetric if `x` and `y` are `d`-neighbours for any positive rational `d`

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  (x y : A)
  where

  is-indistinguishable-prop-Premetric : Prop l2
  is-indistinguishable-prop-Premetric =
    Π-Prop ℚ⁺ (λ (d : ℚ⁺) → B d x y)

  is-indistinguishable-Premetric : UU l2
  is-indistinguishable-Premetric =
    type-Prop is-indistinguishable-prop-Premetric

  is-prop-is-indistinguishable-Premetric :
    is-prop is-indistinguishable-Premetric
  is-prop-is-indistinguishable-Premetric =
    is-prop-type-Prop is-indistinguishable-prop-Premetric
```

> ### Two points `x` and `y` are separated in a premetric if there exists some positive rational `d` such that `x` and `y` are not `d`-neighbours

Reviewer: change header + add definition as prose

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  (x y : A)
  where

  is-separated-pt-prop-Premetric : Prop l2
  is-separated-pt-prop-Premetric =
    ∃ ℚ⁺ (λ d → neg-Prop (B d x y))

  is-separated-pt-Premetric : UU l2
  is-separated-pt-Premetric =
    type-Prop is-separated-pt-prop-Premetric

  is-prop-is-separated-pt-Premetric :
    is-prop is-separated-pt-Premetric
  is-prop-is-separated-pt-Premetric =
    is-prop-type-Prop is-separated-pt-prop-Premetric
```

> ### A premetric is reflexive if any element is indistinguishable from itself

Reviewer: change header  + add definition prose

Author: it is now in a separate module

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  where

  is-reflexive-prop-Premetric : Prop (l1 ⊔ l2)
  is-reflexive-prop-Premetric =
    Π-Prop ℚ⁺ (is-reflexive-prop-Relation-Prop ∘ B)

  is-reflexive-Premetric : UU (l1 ⊔ l2)
  is-reflexive-Premetric = type-Prop is-reflexive-prop-Premetric

  is-prop-is-reflexive-Premetric : is-prop is-reflexive-Premetric
  is-prop-is-reflexive-Premetric =
    is-prop-type-Prop is-reflexive-prop-Premetric
```

> ### A premetric is symmetric if `d`-neighborhoods are symmetric for all positive rational number `d`

Reviewer: change header + add definition as prose

Author: it is now in a separate module

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  where

  is-symmetric-prop-Premetric : Prop (l1 ⊔ l2)
  is-symmetric-prop-Premetric =
    Π-Prop ℚ⁺ (is-symmetric-prop-Relation-Prop ∘ B)

  is-symmetric-Premetric : UU (l1 ⊔ l2)
  is-symmetric-Premetric = type-Prop is-symmetric-prop-Premetric

  is-prop-is-symmetric-Premetric : is-prop is-symmetric-Premetric
  is-prop-is-symmetric-Premetric =
    is-prop-type-Prop is-symmetric-prop-Premetric
```

> ### A premetric is local if indistinguishability has propositional fibers

Reviewer: change header + add definition as prose

Author: it is now in a separate module

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  where

  is-local-prop-Premetric : Prop (l1 ⊔ l2)
  is-local-prop-Premetric =
    Π-Prop A (is-prop-Prop ∘ Σ A ∘ is-indistinguishable-Premetric B)

  is-local-Premetric : UU (l1 ⊔ l2)
  is-local-Premetric = type-Prop is-local-prop-Premetric

  is-prop-is-local-Premetric : is-prop is-local-Premetric
  is-prop-is-local-Premetric = is-prop-type-Prop is-local-prop-Premetric
```

> ### A premetric is tight if any two indistinguishable elements are equal

Author: it is now in a separate module

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  where

  is-tight-Premetric : UU (l1 ⊔ l2)
  is-tight-Premetric =
    (x y : A) → is-indistinguishable-Premetric B x y → x ＝ y
```

> ### A premetric is monotonic if any `d₁`-neighborhoods are `d₂`-neighborhoods for `d₁ < d₂`

Author: it is now in a separate module

```agda
module _
  {l1 l2 : Level} {A : UU l1} (B : Premetric l2 A)
  where

  is-monotonic-prop-Premetric : Prop (l1 ⊔ l2)
  is-monotonic-prop-Premetric =
    Π-Prop
      ( A)
      ( λ x →
        ( Π-Prop
          ( A)
          ( λ y →
            ( Π-Prop
              ( ℚ⁺)
              ( λ d₁ →
                ( Π-Prop
                  ( ℚ⁺)
                  ( λ d₂ →
                    ( Π-Prop
                      ( le-ℚ⁺ d₁ d₂)
                      ( λ H →
                        hom-Prop (B d₁ x y) (B d₂ x y))))))))))

  is-monotonic-Premetric : UU (l1 ⊔ l2)
  is-monotonic-Premetric = type-Prop is-monotonic-prop-Premetric

  is-prop-is-monotonic-Premetric : is-prop is-monotonic-Premetric
  is-prop-is-monotonic-Premetric = is-prop-type-Prop is-monotonic-prop-Premetric
```

> ### A premetric is triangular if it is additively transitive

Reviewer: change header + add definition as prose

Author: it is now in a separate module

This follows Definition 4.5.2 from {{#cite Booij2020PhD}}.

In the classical interpretation, for a premetric `B` on a type `A` and some
positive rational number `d : ℚ⁺`, `B d x y` holds for some points `(x y : A)`

> if the distance between `x` and `y` is bounded above by `d`.

Reviewer: The interpretation isn't really classical, though, right? This is how the reader _should_ interpret the structure

> ### Closeness relation in a premetric

Reviewer: I think I'd reserve the term "closeness" for closeness structures. In that context, closeness should measure how close two things are, not how far away they are.

Author: I'm not sure what you mean. This _does_ measure how close two elements are: if `B d x y` `dist x y ≤ d` so `x` and `y` are `d`-close.

There's actually another equivalence relation hidden there (under good assumptions on the premetric) given by
```
is-neighbor x y = ∃ ℚ⁺ (is-upper-bound-dist-Premetric x y)
 ```
that might be worth defining.

Reviewer: As noted before, "`d`-close" is probably a bad term, and what I mean is that a closeness value of ∞ should mean that two elements are indistinguishable

Reviewer: One of the benefits of the concept of closeness as opposed to distance is that one does not need to assign fractional values to the closeness value of two points.

Author: so 
```
## Neighborhood relation in a premetric
```
maybe?

Reviewer: Isn't a neighborhood relation what a premetric _is_? Do you mean the neighborhood relation associated to a positive fractional value?

Author: well, kind of, but the _premetric_ takes value in `Prop l2` while the `neighborhood-Premetric` relations take value in `UU l2` .

Reviewer: Alright, neighborhood relation associated to an positive fractional value sounds reasonable

Two points `x` and `y` in a type `A` are in a
{{#concept "`d`-neighborhood" Disambiguation="in a premetric" Agda=neighbourhood-Premetric}}
in a premetric `B` for some positive rational numvber `d` if `B d x y` holds.

In this case, `d` is called an
{{#concept "upper bound" Disambiguation="on the distance in a premetric" Agda=is-upper-bound-dist-Premetric}}

> on the distance between `x` and `y` in the premetric `B`.

Reviewer: ```suggestion
Two points `x` and `y` in a type `A` are in a
{{#concept "`d`-neighborhood" Disambiguation="in a premetric" Agda=neighbourhood-Premetric}}
in a [premetric](metric-spaces.premetric-structures.md) `B` for some [positive rational number](elementary-number-theory.positive-rational-numbers.md) `d` if `B d x y` holds.
In this case, we understand `d` to be an
{{#concept "upper bound" Disambiguation="on the distance in a premetric" Agda=is-upper-bound-dist-Premetric}}
on the distance between `x` and `y` with respect to the premetric `B`.
```

Reviewer: Something like this is concise and suggests precisely how the scenario should be interpreted.
